### ISSUE TITLE
Slow dataset prepare

### ISSUE THREAD

**Issue by tcapelle** [huggingface/trl#1299]
I am running this simple script to continue pretraining a model using `SFTTrainer` but the `prepare_datalaoder` method is taking forever...
Any tricks to make this faster?
<img width="483" alt="image" src="https://github.com/huggingface/trl/assets/18441985/5c4e7671-0e91-401f-b0f7-5e912bf18ec0">

The code to reproduce the issue is here:
```python
from types import SimpleNamespace
from datasets import load_dataset
from trl import SFTTrainer

DATASET_NAME = "togethercomputer/RedPajama-Data-1T-Sample"
MODEL_ID = "mistralai/Mistral-7B-v0.1"

config = SimpleNamespace(
    max_seq_length=512,
    seed=42,
)
train_ds = load_dataset(DATASET_NAME)["train"].shuffle(seed=config.seed)

trainer = SFTTrainer(
    MODEL_ID,
    train_dataset=train_ds,
    dataset_text_field="text",
    max_seq_length=config.max_seq_length,
    packing=True,
)
```

I know this is packing the data and making a new dataset under the hood, just want some recommendation best practices on how to deal with fixed length sequences for Causal pre-training (or post-training). Also, maybe a way to paralellize this task?

**Comment by tcapelle** [comment]
I think my issue is coming from the shuffle...

**Comment by younesbelkada** [comment]
Are you in multi-GPU env by any chance? we recently merged https://github.com/huggingface/trl/pull/1255 on main to speed up dataset preparation on multi-GPU envs

**Comment by glorgao** [comment]
@tcapelle Hi, Did you reslove this issue? I encounter the same with the updated trl. After carefully check, I think the data generating is called by the sft trainer rather than the pre-prepare nor the shuffle.

**Comment by qgallouedec** [comment]
This one is pretty old. If you don't get your answer, feel free to open a new issue with everything needed to reproduce

**Comment by glorgao** [comment]
> @tcapelle Hi, Did you reslove this issue? I encounter the same with the updated trl. After carefully check, I think the data generating is called by the sft trainer rather than the pre-prepare nor the shuffle.

FYI
```
import torch.distributed as dist
from datetime import timedelta

dist.init_process_group(backend='nccl', init_method='env://', timeout=timedelta(hours=3))

```
The root of this timeout issue is:
1. preparing a large dataset takes a lot of time (> 30 mins)
2. NCCL by default set a timeout for 30 mins for almost all reduce operator.

So, you may avoid this issue by reducing the size of the dataset or enlarge the timeout period by the above code.
